File: glide/gestures/touchproof_detector.py
"""Main TouchProof detector with state management."""

from __future__ import annotations

from collections import deque

import cv2
import numpy as np

from glide.core.config_models import TouchProofConfig
from glide.core.types import GateState, Landmark
from glide.features.alignment import HandAligner
from glide.gestures.touchproof_scoring import (
    compute_correlation,
    get_adaptive_weights,
    score_angle,
    score_angle_adjusted,
    score_proximity,
    score_proximity_adjusted,
    score_visibility,
)
from glide.gestures.touchproof_signals import MicroFlowTracker, TouchProofSignals


class TouchProofDetector:
    """Multi-signal fusion for robust fingertip contact detection."""

    def __init__(self, config: TouchProofConfig):
        self.config = config
        self.aligner = HandAligner()

        # State tracking
        self.state = GateState.UNARMED
        self._enter_counter = 0
        self._exit_counter = 0

        # Velocity tracking for correlation
        self._idx_positions: deque[tuple[float, float]] = deque(
            maxlen=config.correlation_frames + 1
        )
        self._mid_positions: deque[tuple[float, float]] = deque(
            maxlen=config.correlation_frames + 1
        )

        # Previous frame for velocity
        self._last_update_ms = 0

        # EMA smoothing for volatile signals
        self._proximity_ema: float | None = None
        self._angle_ema: float | None = None  # Added angle smoothing for laptop cameras

        # Baseline tracking for adaptive proximity
        self._baseline_close: float | None = None  # Typical distance when close
        self._baseline_far: float | None = None  # Typical distance when far
        self._baseline_alpha = config.baseline_learning_rate  # From config

        # Micro-flow tracker
        self.flow_tracker = MicroFlowTracker(
            window_frames=config.mfc_window_frames, patch_size=config.mfc_patch_size
        )

        # Cache for expensive computations
        self._last_mfc_score = 0.5

    def update(
        self, landmarks: list[Landmark], frame_bgr: np.ndarray, image_width: int, image_height: int
    ) -> TouchProofSignals:
        """
        Update touch detection with new frame data.

        Args:
            landmarks: Hand landmarks from MediaPipe
            frame_bgr: Current camera frame
            image_width: Frame width
            image_height: Frame height

        Returns:
            TouchProofSignals with all detection signals and final decision
        """
        # Update hand alignment
        if not self.aligner.update(landmarks, image_width, image_height):
            return self._empty_signals()

        # Get fingertip info
        if len(landmarks) < 21:
            return self._empty_signals()

        index_tip = landmarks[8]
        middle_tip = landmarks[12]

        # 1. PROXIMITY SIGNAL
        if self.config.proximity_mode == "logarithmic":
            proximity_norm = self.aligner.get_normalized_distance_log(landmarks)
        else:
            proximity_norm = self.aligner.get_normalized_distance(landmarks)

        # HARD CAP: If fingers are too far apart, no need to compute anything else
        if proximity_norm > self.config.proximity_hard_cap:
            return TouchProofSignals(
                proximity_score=0.0,
                angle_score=0.0,
                correlation_score=0.0,
                visibility_score=0.0,
                mfc_score=0.0,
                distance_factor=self.aligner.get_hand_distance_factor(),
                fused_score=0.0,
                is_touching=False,
            )

        proximity_score_raw = score_proximity(
            proximity_norm, self.config.proximity_enter, self.config.proximity_exit
        )

        # Apply EMA smoothing if enabled
        if self.config.smooth_proximity:
            if self._proximity_ema is None:
                self._proximity_ema = proximity_score_raw
            else:
                self._proximity_ema = (
                    self.config.ema_alpha * proximity_score_raw
                    + (1 - self.config.ema_alpha) * self._proximity_ema
                )
            proximity_score = self._proximity_ema
        else:
            proximity_score = proximity_score_raw

        # 2. ANGLE SIGNAL
        angle_deg = self.aligner.get_fingertip_angle(landmarks)

        # HARD CAP: If fingers are pointing in very different directions
        if angle_deg > self.config.angle_hard_cap_deg:
            return TouchProofSignals(
                proximity_score=proximity_score,
                angle_score=0.0,
                correlation_score=0.0,
                visibility_score=0.0,
                mfc_score=0.0,
                distance_factor=self.aligner.get_hand_distance_factor(),
                fused_score=0.0,
                is_touching=False,
            )

        # Apply angle smoothing for laptop camera stability
        angle_alpha = 0.2  # Faster response than proximity (0.3)
        if self._angle_ema is None:
            self._angle_ema = angle_deg
        else:
            self._angle_ema = angle_alpha * angle_deg + (1 - angle_alpha) * self._angle_ema

        angle_score = score_angle(
            self._angle_ema, self.config.angle_enter_deg, self.config.angle_exit_deg
        )

        # 3. MOTION CORRELATION SIGNAL
        # Update position buffers
        idx_aligned = self.aligner.to_hand_aligned(index_tip.x, index_tip.y)
        mid_aligned = self.aligner.to_hand_aligned(middle_tip.x, middle_tip.y)
        self._idx_positions.append(idx_aligned)
        self._mid_positions.append(mid_aligned)

        correlation_score = compute_correlation(
            self._idx_positions,
            self._mid_positions,
            self.config.correlation_frames,
            self.config.correlation_min,
        )

        # 4. VISIBILITY/OCCLUSION SIGNAL
        visibility_score = score_visibility(
            index_tip, middle_tip, self.config.visibility_asymmetry_min
        )

        # 5. Get fingertip pixel coordinates (needed for MFC)
        index_px, middle_px = self.aligner.get_fingertip_pixels(landmarks)

        # 6. Get distance factor for adaptive fusion
        distance_factor = self.aligner.get_hand_distance_factor()

        # 6b. Update baseline distances for adaptive proximity
        self._update_baseline(proximity_norm, distance_factor)

        # 7. Compute initial fused score for conditional logic
        initial_fused = 0.7 * proximity_score + 0.3 * angle_score

        # 8. MFC (Micro-Flow Cohesion) - expanded band for laptop cameras
        if (
            self.state == GateState.READY
            or (0.40 <= initial_fused <= 0.70)  # Expanded uncertainty band
            or distance_factor < 0.3
        ):  # Always compute when very close
            # Convert to grayscale for optical flow
            gray = cv2.cvtColor(frame_bgr, cv2.COLOR_BGR2GRAY)
            mfc_score = self.flow_tracker.update(gray, index_px, middle_px)
            self._last_mfc_score = mfc_score
        else:
            mfc_score = self._last_mfc_score

        # 9. DISTANCE-AWARE FUSION
        weights = get_adaptive_weights(distance_factor)

        # Get baseline for adaptive mode
        baseline = self._get_baseline_distance(distance_factor)

        # Recalculate scores with adjusted thresholds
        proximity_score_adj = score_proximity_adjusted(
            proximity_norm, distance_factor, self.config, baseline
        )
        angle_score_adj = score_angle_adjusted(angle_deg, distance_factor, self.config)

        # Fuse all signals with adaptive weights
        fused_score = (
            weights["proximity"] * proximity_score_adj
            + weights["angle"] * angle_score_adj
            + weights["mfc"] * mfc_score
            + weights["occlusion"] * visibility_score
        )

        # STATE MACHINE with hysteresis
        is_touching = self._update_state(fused_score)

        return TouchProofSignals(
            proximity_score=proximity_score_adj,
            angle_score=angle_score_adj,
            correlation_score=correlation_score,
            visibility_score=visibility_score,
            mfc_score=mfc_score,
            distance_factor=distance_factor,
            fused_score=fused_score,
            is_touching=is_touching,
        )

    def _update_state(self, fused_score: float) -> bool:
        """Update state machine with hysteresis."""
        # Simple threshold with hysteresis
        if self.state == GateState.UNARMED:
            if fused_score > self.config.fused_enter_threshold:
                self._enter_counter += 1
                if self._enter_counter >= self.config.frames_to_enter:
                    self.state = GateState.READY
                    self._enter_counter = 0
                    return True
            else:
                self._enter_counter = 0
            return False

        if self.state == GateState.READY:
            if fused_score < self.config.fused_exit_threshold:
                self._exit_counter += 1
                if self._exit_counter >= self.config.frames_to_exit:
                    self.state = GateState.UNARMED
                    self._exit_counter = 0
                    return False
            else:
                self._exit_counter = 0
            return True

        return False

    def _empty_signals(self) -> TouchProofSignals:
        """Return empty signals when detection fails."""
        return TouchProofSignals(
            proximity_score=0.0,
            angle_score=0.0,
            correlation_score=0.0,
            visibility_score=0.0,
            mfc_score=0.0,
            distance_factor=0.5,
            fused_score=0.0,
            is_touching=False,
        )

    def _update_baseline(self, distance_norm: float, distance_factor: float) -> None:
        """Update baseline distances for different hand distances."""
        # Only update when not touching (to learn normal separation)
        if self.state == GateState.UNARMED:
            if distance_factor < 0.3:  # Close
                if self._baseline_close is None:
                    self._baseline_close = distance_norm
                else:
                    self._baseline_close = (
                        self._baseline_alpha * distance_norm
                        + (1 - self._baseline_alpha) * self._baseline_close
                    )
            elif distance_factor > 0.7:  # Far
                if self._baseline_far is None:
                    self._baseline_far = distance_norm
                else:
                    self._baseline_far = (
                        self._baseline_alpha * distance_norm
                        + (1 - self._baseline_alpha) * self._baseline_far
                    )

    def _get_baseline_distance(self, distance_factor: float) -> float | None:
        """Get expected baseline distance for current hand distance."""
        if self._baseline_close is None or self._baseline_far is None:
            return None

        if distance_factor < 0.3:
            return self._baseline_close
        if distance_factor > 0.7:
            return self._baseline_far
        # Linear interpolation
        t = (distance_factor - 0.3) / 0.4
        return self._baseline_close * (1 - t) + self._baseline_far * t

File: apps/web-ui/src/gestures/touchproof-detector.ts
/**
 * TouchProof detector — multi-signal fusion for fingertip contact detection.
 * Ported from glide/gestures/touchproof_detector.py
 *
 * MFC (optical flow) is excluded: mfcScore hardcoded to 0.5 (neutral).
 */

import type { TouchProofConfig } from '../core/config';
import { GateState } from '../core/types';
import type { Landmark } from '../core/types';
import { BoundedDeque } from '../core/collections';
import { HandAligner } from '../features/alignment';
import {
  scoreProximity,
  scoreAngle,
  scoreVisibility,
  computeCorrelation,
  scoreProximityAdjusted,
  scoreAngleAdjusted,
  getAdaptiveWeights,
} from './touchproof-scoring';
import type { TouchProofSignals } from './touchproof-signals';
import { emptySignals } from './touchproof-signals';

export class TouchProofDetector {
  config: TouchProofConfig;
  aligner: HandAligner;
  state: GateState = GateState.UNARMED;

  private _enterCounter = 0;
  private _exitCounter = 0;

  private _idxPositions: BoundedDeque<[number, number]>;
  private _midPositions: BoundedDeque<[number, number]>;

  private _proximityEma: number | null = null;
  private _angleEma: number | null = null;

  private _baselineClose: number | null = null;
  private _baselineFar: number | null = null;
  private _baselineAlpha: number;

  /** Hardcoded MFC score (no optical flow in browser). */
  private _lastMfcScore = 0.5;

  constructor(config: TouchProofConfig) {
    this.config = config;
    this.aligner = new HandAligner();
    this._idxPositions = new BoundedDeque(config.correlationFrames + 1);
    this._midPositions = new BoundedDeque(config.correlationFrames + 1);
    this._baselineAlpha = config.baselineLearningRate;
  }

  /**
   * Update touch detection with new landmarks.
   * Note: no frame_bgr param (MFC excluded).
   */
  update(
    landmarks: Landmark[],
    imageWidth: number,
    imageHeight: number,
  ): TouchProofSignals {
    if (!this.aligner.update(landmarks, imageWidth, imageHeight)) {
      return emptySignals();
    }

    if (landmarks.length < 21) return emptySignals();

    const indexTip = landmarks[8]!;
    const middleTip = landmarks[12]!;

    // 1. PROXIMITY
    const proximityNorm =
      this.config.proximityMode === 'logarithmic'
        ? this.aligner.getNormalizedDistanceLog(landmarks)
        : this.aligner.getNormalizedDistance(landmarks);

    // Hard cap
    if (proximityNorm > this.config.proximityHardCap) {
      return {
        proximityScore: 0,
        angleScore: 0,
        correlationScore: 0,
        visibilityScore: 0,
        mfcScore: 0,
        distanceFactor: this.aligner.getHandDistanceFactor(),
        fusedScore: 0,
        isTouching: false,
      };
    }

    const proximityScoreRaw = scoreProximity(
      proximityNorm,
      this.config.proximityEnter,
      this.config.proximityExit,
    );

    let proximityScore: number;
    if (this.config.smoothProximity) {
      if (this._proximityEma == null) {
        this._proximityEma = proximityScoreRaw;
      } else {
        this._proximityEma =
          this.config.emaAlpha * proximityScoreRaw +
          (1 - this.config.emaAlpha) * this._proximityEma;
      }
      proximityScore = this._proximityEma;
    } else {
      proximityScore = proximityScoreRaw;
    }

    // 2. ANGLE
    const angleDeg = this.aligner.getFingertipAngle(landmarks);

    if (angleDeg > this.config.angleHardCapDeg) {
      return {
        proximityScore,
        angleScore: 0,
        correlationScore: 0,
        visibilityScore: 0,
        mfcScore: 0,
        distanceFactor: this.aligner.getHandDistanceFactor(),
        fusedScore: 0,
        isTouching: false,
      };
    }

    // Angle smoothing
    const angleAlpha = 0.2;
    if (this._angleEma == null) {
      this._angleEma = angleDeg;
    } else {
      this._angleEma = angleAlpha * angleDeg + (1 - angleAlpha) * this._angleEma;
    }

    // Raw angle score computed for consistency with Python;
    // the adjusted version (scoreAngleAdjusted) is used in fusion below.
    scoreAngle(
      this._angleEma,
      this.config.angleEnterDeg,
      this.config.angleExitDeg,
    );

    // 3. CORRELATION
    const idxAligned = this.aligner.toHandAligned(indexTip.x, indexTip.y);
    const midAligned = this.aligner.toHandAligned(middleTip.x, middleTip.y);
    this._idxPositions.push(idxAligned);
    this._midPositions.push(midAligned);

    const correlationScore = computeCorrelation(
      this._idxPositions.toArray(),
      this._midPositions.toArray(),
      this.config.correlationFrames,
      this.config.correlationMin,
    );

    // 4. VISIBILITY
    const visibilityScore = scoreVisibility(
      indexTip,
      middleTip,
      this.config.visibilityAsymmetryMin,
    );

    // 5. DISTANCE FACTOR
    const distanceFactor = this.aligner.getHandDistanceFactor();

    // 6. BASELINE UPDATE
    this._updateBaseline(proximityNorm, distanceFactor);

    // 7. MFC (hardcoded — no optical flow in browser)
    const mfcScore = this._lastMfcScore;

    // 8. DISTANCE-AWARE FUSION
    const weights = getAdaptiveWeights(distanceFactor);
    const baseline = this._getBaselineDistance(distanceFactor);

    const proximityScoreAdj = scoreProximityAdjusted(
      proximityNorm,
      distanceFactor,
      this.config,
      baseline,
    );
    const angleScoreAdj = scoreAngleAdjusted(angleDeg, distanceFactor, this.config);

    const fusedScore =
      weights.proximity * proximityScoreAdj +
      weights.angle * angleScoreAdj +
      weights.mfc * mfcScore +
      weights.occlusion * visibilityScore;

    // STATE MACHINE
    const isTouching = this._updateState(fusedScore);

    return {
      proximityScore: proximityScoreAdj,
      angleScore: angleScoreAdj,
      correlationScore,
      visibilityScore,
      mfcScore,
      distanceFactor,
      fusedScore,
      isTouching,
    };
  }

  private _updateState(fusedScore: number): boolean {
    if (this.state === GateState.UNARMED) {
      if (fusedScore > this.config.fusedEnterThreshold) {
        this._enterCounter++;
        if (this._enterCounter >= this.config.framesToEnter) {
          this.state = GateState.READY;
          this._enterCounter = 0;
          return true;
        }
      } else {
        this._enterCounter = 0;
      }
      return false;
    }

    if (this.state === GateState.READY) {
      if (fusedScore < this.config.fusedExitThreshold) {
        this._exitCounter++;
        if (this._exitCounter >= this.config.framesToExit) {
          this.state = GateState.UNARMED;
          this._exitCounter = 0;
          return false;
        }
      } else {
        this._exitCounter = 0;
      }
      return true;
    }

    return false;
  }

  private _updateBaseline(distanceNorm: number, distanceFactor: number): void {
    if (this.state !== GateState.UNARMED) return;

    if (distanceFactor < 0.3) {
      if (this._baselineClose == null) {
        this._baselineClose = distanceNorm;
      } else {
        this._baselineClose =
          this._baselineAlpha * distanceNorm + (1 - this._baselineAlpha) * this._baselineClose;
      }
    } else if (distanceFactor > 0.7) {
      if (this._baselineFar == null) {
        this._baselineFar = distanceNorm;
      } else {
        this._baselineFar =
          this._baselineAlpha * distanceNorm + (1 - this._baselineAlpha) * this._baselineFar;
      }
    }
  }

  private _getBaselineDistance(distanceFactor: number): number | null {
    if (this._baselineClose == null || this._baselineFar == null) return null;

    if (distanceFactor < 0.3) return this._baselineClose;
    if (distanceFactor > 0.7) return this._baselineFar;

    const t = (distanceFactor - 0.3) / 0.4;
    return this._baselineClose * (1 - t) + this._baselineFar * t;
  }
}
